### PR TITLE
Initial NTLM Implementation

### PR DESCRIPTION
**Pull request by rvazarkar**
## Description
This is the initial implementation of the new NTLM modelling, courtesy of @leechristensen . 

## Motivation and Context

https://specterops.atlassian.net/browse/BED-5113

## Types of changes

-   [ ] Chore (a change that does not modify the application functionality)
-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

-   [ ] Documentation updates are needed, and have been made accordingly.
-   [ ] I have added and/or updated tests to cover my changes.
-   [ ] All new and existing tests passed.
-   [ ] My changes include a database migration.

### CODE REVIEW

File: src/CommonLib/Ntlm/HttpNtlmAuthenticationService.cs [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+    public HttpNtlmAuthenticationService(ILogger logger, IHttpClientFactory httpClientFactory)
+    {
+        _logger = logger;
+        _httpClientFactory = httpClientFactory;
+    }
+
+    public async Task EnsureRequiresAuth(bool enableSigning, bool? useBadChannelBindings)
+    {
+        if (Url == null)
```
Comment (mvlipka) [inline src/CommonLib/Ntlm/HttpNtlmAuthenticationService.cs]: It looks like the `Url` property is only used on this method while being passed in to our other methods, might be worth moving this to a required argument rather than a property

File: src/CommonLib/Ntlm/HttpTransport.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+using System;
+using System.Linq;
+using System.Net.Http;
+using System.Threading.Tasks;
+
+namespace SharpHoundCommonLib.Ntlm;
+
+public class HttpTransport : INtlmTransport
+{
+    private readonly ILogger _logger;
```
Comment (mvlipka) [inline src/CommonLib/Ntlm/HttpTransport.cs:14]: `_logger` doesn't seem to be used, might not be needed, or potentially was forgot to be used

File: src/CommonLib/Ntlm/HttpTransport.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+        var response = await _httpClient.SendAsync(requestMessage);
+
+        if (!response.Headers.Contains("WWW-Authenticate"))
+        {
+            throw new InvalidOperationException("No WWW-Authenticate header found in response");
+        }
+
+        var authHeaders = response.Headers.WwwAuthenticate.Where(a => a.Scheme == _authScheme);
+        if (!authHeaders.Any())
```
Comment (mvlipka) [inline src/CommonLib/Ntlm/HttpTransport.cs:38]: Slight performance nit: I think we can use `authHeaders.Count == 0` here

File: src/CommonLib/Ntlm/HttpTransport.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        requestMessage.Headers.Add("Authorization", $"{_authScheme} {messageBase64}");
+
+        var response = await _httpClient.SendAsync(requestMessage);
+
+        if (!response.Headers.Contains("WWW-Authenticate"))
+        {
+            throw new InvalidOperationException("No WWW-Authenticate header found in response");
+        }
+
+        var authHeaders = response.Headers.WwwAuthenticate.Where(a => a.Scheme == _authScheme);
```
Comment (mvlipka) [inline src/CommonLib/Ntlm/HttpTransport.cs]: Slight performance nit: we can convert this to a list to avoid enumerating the elements again, probably not a large enough collection to make a large difference

File: src/CommonLib/Ntlm/LdapNative.cs [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+
+            // Freeing the memory always results in crashes. Is it managed by something else?
+            //if (response != IntPtr.Zero)
+            //{
+            //    var responseBerval = Marshal.PtrToStructure<NativeMethods.berval>(response);
+            //    if (responseBerval.bv_val != IntPtr.Zero)
+            //    {
+            //        Marshal.FreeHGlobal(responseBerval.bv_val);
+            //    }
+            //    Marshal.FreeHGlobal(response);
```
Comment (mvlipka) [inline src/CommonLib/Ntlm/LdapNative.cs:241]: I _think_ that the memory is cleaned up at the end of the PInvoke call since, which could potentially be causing the crash

File: src/CommonLib/Ntlm/NtlmAuthenticationHandler.cs [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+                AuthenticationMethod.NTLM,
+                _targetService,
+                Options.Bindings,
+                Options.Signing,
+                Options.Signing
+            );
+
+        // NEGOTIATE
+        var negotiateMsgBytes = context.Step();
+        //_logger.LogDebug($"NTLM-NEGOTIATE message: {Convert.ToBase64String(negotiateMsgBytes)}");
```
Comment (mvlipka) [inline src/CommonLib/Ntlm/NtlmAuthenticationHandler.cs]: Not sure if intentionally left out, but _logger is no longer used if intentional

We also don't use `_host` or `NtlmAuthPackageName`

File: src/CommonLib/OutputTypes/APIResult.cs [hunk 7]
```diff
@@ -1,8 +1,52 @@
 ﻿namespace SharpHoundCommonLib.OutputTypes
 {
-    public class APIResult
+    public class BaseAPIResult<T>
```
Comment (mvlipka) [inline src/CommonLib/OutputTypes/APIResult.cs]: Unused, can probably remove

File: src/CommonLib/OutputTypes/NtlmSession.cs [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+            return ($"  {TimeCreatedUtc?.ToLocalTime()},{Id},{targetUser},{AccountSid},{SourceHost},,{PackageName}");
+        }
+
+        public static NtlmSession FromLogonEvent(string host, EventRecord evnt)
+        {
+            if (evnt.Id != 4624)
+                throw new ArgumentException("Not a logon event");
+
+            //var subjectUserSid = eventDetail.Properties[0].Value.ToString();
```
Comment (mvlipka) [inline src/CommonLib/OutputTypes/NtlmSession.cs]: Thinking we can remove these

File: src/CommonLib/OutputTypes/NtlmSession.cs [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+        public string? AccountDomain { get; set; } = accountDomain;
+        public string SourceHost { get; set; } = sourceHost;   // The host the auth originated from
+        public string? SourceIp { get; set; } = sourceIp;
+        public string? SourcePort { get; set; } = sourcePort;
+        public string? PackageName { get; set; } = packageName;
+
+        public override string ToString()
+        {
+            var targetUser = AccountDomain + "\\" + AccountName;
+            var source = (SourceIp != null || SourcePort != null) ? $"{SourceIp}:{SourcePort}" : "";
```
Comment (mvlipka) [inline src/CommonLib/OutputTypes/NtlmSession.cs:36]: Unused, maybe was meant for logging?

File: src/CommonLib/OutputTypes/NtlmSession.cs [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        string? accountSid,
+        string accountName,
+        string? accountDomain,
+        string sourceHost,
+        string? sourceIp,
+        string? sourcePort,
+        string packageName
+        )
+    {
+        public string Host { get; set; } = host;   // Hostname the sessions occurred on
```
Comment (mvlipka) [inline src/CommonLib/OutputTypes/NtlmSession.cs]: Unused

File: src/CommonLib/OutputTypes/NtlmSession.cs [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+        {
+            var targetUser = AccountDomain + "\\" + AccountName;
+            var source = (SourceIp != null || SourcePort != null) ? $"{SourceIp}:{SourcePort}" : "";
+
+            return ($"  {TimeCreatedUtc?.ToLocalTime()},{Id},{targetUser},{AccountSid},{SourceHost},,{PackageName}");
+        }
+
+        public static NtlmSession FromLogonEvent(string host, EventRecord evnt)
+        {
+            if (evnt.Id != 4624)
```
Comment (mvlipka) [inline src/CommonLib/OutputTypes/NtlmSession.cs]: Should these magic numbers be moved to an enum? Maybe a constant for this

File: src/CommonLib/Processors/CAEnrollmentProcessor.cs [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+
+        public async Task<IEnumerable<ApiResult<CAEnrollmentEndpoint>>> ScanAsync()
+        {
+            var endpoints = new List<ApiResult<CAEnrollmentEndpoint>>();
+
+            try
+            {
+                var webEnrollmentTask = ScanHttpEndpoint(CAEnrollmentEndpointType.WebEnrollmentApplication);
+                var webServiceTask = ScanHttpEndpoint(CAEnrollmentEndpointType.EnrollmentWebService);
+                // var rpcServiceTask = CheckEnrollmentRpcAsync();
```
Comment (mvlipka) [inline src/CommonLib/Processors/CAEnrollmentProcessor.cs]: Can remove the comments

File: src/CommonLib/Processors/DCLdapProcessor.cs [hunk 13]
```diff
@@ -0,0 +1,291 @@
...
+
+            return ApiResult<bool>.CreateSuccess(accessibleWithoutSigning == false);
+        }
+        catch (Exception ex)
+        {
+            return ApiResult<bool>.CreateError($"CheckIsNtlmSigningRequired failed: {ex}");
+        }
+    }
+
+    // Checks if EPA is enabled. Does so by trying to auth with wrong channel bingings.
```
Comment (mvlipka) [inline src/CommonLib/Processors/DCLdapProcessor.cs]: bingings->bindings

File: src/CommonLib/Processors/DCLdapProcessor.cs [hunk 14]
```diff
@@ -0,0 +1,291 @@
...
+            };
+            var accessibleWithNoBindings = await Authenticate(_ldapSslEndpoint, new LdapAuthOptions()
+            {
+                Signing = false,
+                Bindings = bindings
+            });
+
+            return ApiResult<bool>.CreateSuccess(accessibleWithNoBindings == true);
+
+            //// Second, check if we can connect, but with *INVALID* bindings
```
Comment (mvlipka) [inline src/CommonLib/Processors/DCLdapProcessor.cs]: Looks like we can remove these

File: src/CommonLib/Processors/DCLdapProcessor.cs [hunk 15]
```diff
@@ -0,0 +1,291 @@
...
+        var auth = new NtlmAuthenticationHandler(_log, host, $"LDAP/{host.ToUpper()}")
+        {
+            Options = options
+        };
+        var transport = new LdapTransport(_log, endpoint);
+
+        try
+        {
+            transport.InitializeConnectionAsync(_ldapTimeout);
+            var bytes = (byte[])await auth.PerformNtlmAuthenticationAsync(transport);
```
Comment (mvlipka) [inline src/CommonLib/Processors/DCLdapProcessor.cs]: bytes is unused, we can avoid the extra memory initialization

File: src/CommonLib/Ntlm/LdapNative.cs [hunk 16]
```diff
@@ -0,0 +1,346 @@
...
+            // If anything goes wrong getting the server error,
+            // return null rather than throw since we're already
+            // in error handling code
+            return null;
+        }
+    }
+
+    public void CheckError(int result, string operation)
+    {
+        if (result != 0)
```
Comment (mvlipka) [inline src/CommonLib/Ntlm/LdapNative.cs]: I think we could replace `0` with `LdapErrorCodes.Success`, but not sure if the return of the `ldap_set_option` system call is the same as the out value

File: src/CommonLib/Processors/DCLdapProcessor.cs [hunk 17]
```diff
@@ -0,0 +1,291 @@
...
+        }
+        catch (Exception ex)
+        {
+            _log.LogError("An unhandled error occurred during the LDAP test: {ex}", ex);
+        }
+
+        return false;
+    }
+
+    //private Task<bool> CanAccessWithChannelBindings()
```
Comment (mvlipka) [inline src/CommonLib/Processors/DCLdapProcessor.cs]: Removable

File: src/CommonLib/Processors/EventLogProcessor.cs [hunk 18]
```diff
@@ -0,0 +1,332 @@
...
+         string host,
+         string domain,
+         EventLogCollection collectionType,
+         int numDays = 7,
+         int readEventDelayMs = 5,
+         int readEventTimeoutMs = 10000,
+         NetworkCredential? localAdminCredential = null
+         )
+    {
+        private readonly ILdapUtils _utils = utils;
```
Comment (mvlipka) [inline src/CommonLib/Processors/EventLogProcessor.cs]: Looks like a majority of these private fields are unused

File: src/CommonLib/Processors/EventLogProcessor.cs [hunk 19]
```diff
@@ -0,0 +1,332 @@
...
+        }
+
+        private List<NtlmSession> CollectEvents(string query)
+        {
+            var ntlmSessions = new List<NtlmSession>();
+            var dupCheck = new HashSet<string>();
+
+            using EventLogReader logReader = GetEventLogReader(_host, "Security", query);
+
+            int count = 0;
```
Comment (mvlipka) [inline src/CommonLib/Processors/EventLogProcessor.cs]: We don't seem to do anything with the `count` here

File: src/CommonLib/Processors/WebClientServiceProcessor.cs [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+using System.Threading.Tasks;
+
+namespace SharpHoundCommonLib.Processors
+{
+    public class WebClientServiceProcessor(ILogger log = null)
+    {
+        private readonly ILogger _log = log ?? Logging.LogProvider.CreateLogger("WebClientServiceProcessor");
+
+        // Define constants
+        public const uint MAXIMUM_ALLOWED = 0x02000000;
```
Comment (mvlipka) [inline src/CommonLib/Processors/WebClientServiceProcessor.cs:21]: I believe these match the naming style of the library calls, but not sure we want to copy that naming style into our managed code

File: src/CommonLib/Processors/WebClientServiceProcessor.cs [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+            uint dwDesiredAccess,
+            uint dwShareMode,
+            IntPtr lpSecurityAttributes,
+            uint dwCreationDisposition,
+            uint dwFlagsAndAttributes,
+            IntPtr hTemplateFile);
+
+        public bool TestPathExists(string path)
+        {
+            using SafeFileHandle handle = CreateFile(
```
Comment (mvlipka) [inline src/CommonLib/Processors/WebClientServiceProcessor.cs]: Is there a benefit of this over using `File.Create`?

File: src/CommonLib/ThirdParty/PSOpenAD/Authentication.cs [hunk 22]
```diff
@@ -0,0 +1,515 @@
...
+    }
+
+    public override byte[] Wrap(ReadOnlySpan<byte> data, bool encrypt) => Array.Empty<byte>();
+    public override byte[] Unwrap(ReadOnlySpan<byte> data) => Array.Empty<byte>();
+    public override uint MaxWrapSize(uint outputSize, bool confReq) => 0;
+    public override void Dispose()
+    { }
+}
+
+//internal class GssapiContext : SecurityContext
```
Comment (mvlipka) [inline src/CommonLib/ThirdParty/PSOpenAD/Authentication.cs]: Removable

File: src/CommonLib/ThirdParty/PSOpenAD/Authentication.cs [hunk 23]
```diff
@@ -0,0 +1,515 @@
...
+    private UInt32 _trailerSize = 0;
+    private UInt32 _seqNo = 0;
+
+    public SspiContext(string? username, string? password, AuthenticationMethod method, string target,
+        ChannelBindings? channelBindings, bool integrity, bool confidentiality)
+    {
+        _bindingData = CreateChannelBindings(channelBindings);
+        _targetSpn = target;
+
+        string package = Enum.GetName(typeof(AuthenticationMethod), method);
```
Comment (mvlipka) [inline src/CommonLib/ThirdParty/PSOpenAD/Authentication.cs]: I think `method.ToString()` might be more appropriate here

File: src/CommonLib/ThirdParty/PSOpenAD/Authentication.cs [hunk 24]
```diff
@@ -0,0 +1,515 @@
...
+
+                if (_bindingData != null)
+                {
+                    inputBuffers[idx].cbBuffer = (UInt32)_bindingData.Length;
+                    inputBuffers[idx].BufferType = (UInt32)SecBufferType.SECBUFFER_CHANNEL_BINDINGS;
+                    inputBuffers[idx].pvBuffer = (IntPtr)cbBuffer;
+                }
+
+                SspiSecContext context = SSPI.InitializeSecurityContext(_credential, _context, _targetSpn, _flags,
+                    TargetDataRep.SECURITY_NATIVE_DREP, inputBuffers, new[] { SecBufferType.SECBUFFER_TOKEN, });
```
Comment (mvlipka) [inline src/CommonLib/ThirdParty/PSOpenAD/Authentication.cs:220]: Nit: we can replace `new [] {}` with `[SecBufferType.SECBUFFER_TOKEN]`

File: src/CommonLib/ThirdParty/PSOpenAD/Authentication.cs [hunk 25]
```diff
@@ -0,0 +1,515 @@
...
+                        offset += (int)buffers[0].cbBuffer;
+                    }
+
+                    Marshal.Copy((IntPtr)dataPtr, wrapped, offset, (int)buffers[1].cbBuffer);
+                    offset += (int)buffers[1].cbBuffer;
+
+                    if (buffers[2].cbBuffer > 0)
+                    {
+                        Buffer.BlockCopy(padding, 0, wrapped, offset, (int)buffers[2].cbBuffer);
+                        offset += (int)buffers[2].cbBuffer;
```
Comment (mvlipka) [inline src/CommonLib/ThirdParty/PSOpenAD/Authentication.cs:292]: We can remove since we no longer use the `offset` after this if block, might have been kept for future changes, though?

File: src/CommonLib/ThirdParty/PSOpenAD/SSPI.cs [hunk 26]
```diff
@@ -0,0 +1,676 @@
...
+using System.ComponentModel;
+using System.Runtime.InteropServices;
+using System.Security.Authentication;
+
+namespace SharpHoundCommonLib.ThirdParty.PSOpenAD;
+
+internal static partial class Helpers
+{
+    [StructLayout(LayoutKind.Sequential)]
+    public struct SEC_CHANNEL_BINDINGS
```
Comment (mvlipka) [inline src/CommonLib/ThirdParty/PSOpenAD/SSPI.cs:15]: Pretty sure these are the names from the shared library, but we should follow our managed code's styling

File: src/CommonLib/ThirdParty/PSOpenAD/SSPI.cs [hunk 27]
```diff
@@ -0,0 +1,676 @@
...
+        OutputBuffers = outputBuffers;
+        Expiry = expiry;
+        Flags = flags;
+        MoreNeeded = moreNeeded;
+    }
+}
+
+internal static class SSPI
+{
+    private const Int32 SEC_I_CONTINUE_NEEDED = 0x00090312;
```
Comment (mvlipka) [inline src/CommonLib/ThirdParty/PSOpenAD/SSPI.cs:126]: Same here with matching C#'s styles

File: src/CommonLib/Ntlm/LdapNative.cs [hunk 28]
```diff
@@ -0,0 +1,346 @@
...
+
+    [DllImport("Wldap32.dll", EntryPoint = "ldap_sasl_bind_s", CallingConvention = CallingConvention.Cdecl)]
+    public static extern int ldap_sasl_bind_s(
+        SafeLdapHandle ld,
+        string dn,
+        string mechanism,
+        IntPtr cred,
+        IntPtr serverctrls,
+        IntPtr clientctrls,
+        out IntPtr msgidp);
```
Comment (mvlipka) [inline src/CommonLib/Ntlm/LdapNative.cs:297]: Nit: this name doesn't match the output defined in the library
https://learn.microsoft.com/en-us/windows/win32/api/winldap/nf-winldap-ldap_sasl_bind_sa

File: src/CommonLib/ThirdParty/PSOpenAD/SSPI.cs [hunk 29]
```diff
@@ -0,0 +1,676 @@
...
+    SECPKG_CRED_BOTH = 0x00000003,
+    SECPKG_CRED_DEFAULT = 0x00000004,
+    SECPKG_CRED_AUTOLOGON_RESTRICTED = 0x00000010,
+    SECPKG_CRED_PROCESS_POLICY_ONLY = 0x00000020,
+}
+
+[Flags]
+internal enum InitiatorContextRequestFlags : uint
+{
+    ISC_REQ_DELEGATE = 0x00000001,
```
Comment (mvlipka) [inline src/CommonLib/ThirdParty/PSOpenAD/SSPI.cs:481]: Naming in these enums as well

**Review by mvlipka — CHANGES_REQUESTED** [review]
Looking good, mostly style comments and naming conventions that I think we can improve on

File: src/CommonLib/LdapProducerQueryGenerator.cs [hunk 30]
```diff
@@ -46,7 +46,7 @@ public static GeneratedLdapParameters GenerateDefaultPartitionParameters(Collect
             if (methods.HasFlag(CollectionMethod.SPNTargets))
                 properties.AddRange(CommonProperties.SPNTargetProps);
 
-            if (methods.HasFlag(CollectionMethod.DCRegistry))
+            if (methods.HasFlag(CollectionMethod.DCRegistry) || methods.HasFlag(CollectionMethod.LdapServices))
```
Comment (rvazarkar) [inline src/CommonLib/LdapProducerQueryGenerator.cs]: Does this need WebClientService/SMBInfo/EventLogs added to it?

File: src/CommonLib/Ntlm/LdapNative.cs (hunk 28 above)
Comment (rvazarkar) [reply src/CommonLib/Ntlm/LdapNative.cs:297]: Unfortunately, pinvoke is weird, this is normal

File: src/CommonLib/Processors/WebClientServiceProcessor.cs (hunk 21 above)
Comment (rvazarkar) [reply src/CommonLib/Processors/WebClientServiceProcessor.cs]: Good question, @leechristensen can you clarify?

Comment (leechristensen) [reply src/CommonLib/Processors/WebClientServiceProcessor.cs]: I believe it was because File.Create always throws an exception for the pipe whereas we can get a success (and more granular error checking) with lower-level win32 APIs. I believe the error handling seemed simpler for Win32 APIs vs dealing with the various exceptions that .NET's File class may throw. But not tied to it either way.

File: src/CommonLib/Ntlm/HttpTransport.cs (hunk 2 above)
Comment (rvazarkar) [reply src/CommonLib/Ntlm/HttpTransport.cs:14]: I'm ok with leaving unused loggers for now, it lets us quickly add logging later

File: src/CommonLib/OutputTypes/NtlmSession.cs (hunk 8 above)
Comment (rvazarkar) [reply src/CommonLib/OutputTypes/NtlmSession.cs]: We can, but I think they serve as documentation for that structure thats coming back, which isn't bad to have in case we need the info later

File: src/CommonLib/Processors/WebClientServiceProcessor.cs (hunk 20 above)
Comment (rvazarkar) [reply src/CommonLib/Processors/WebClientServiceProcessor.cs:21]: I'm going to leave it as is to match the well defined consts

File: src/CommonLib/Processors/WebClientServiceProcessor.cs (hunk 21 above)
Comment (rvazarkar) [reply src/CommonLib/Processors/WebClientServiceProcessor.cs]: Adding a comment and resolving

File: src/CommonLib/ThirdParty/PSOpenAD/Authentication.cs (hunk 25 above)
Comment (rvazarkar) [reply src/CommonLib/ThirdParty/PSOpenAD/Authentication.cs:292]: I don't know the intention for that so I'm going to leave it as is for now, we can refactor later

**Comment by rvazarkar** [comment]
recheck

File: src/CommonLib/OutputTypes/NtlmSession.cs (hunk 9 above)
Comment (rvazarkar) [reply src/CommonLib/OutputTypes/NtlmSession.cs:36]: Added into the log statement

**Review by mvlipka — APPROVED** [review]

File: src/CommonLib/LdapUtils.cs [hunk 31]
```diff
@@ -1203,6 +1203,15 @@ await utils.GetDomainNameFromSid(objectIdentifier) is (true, var domainName)) {
                 return (true, res);
             }
 
+            res.ObjectType = await ComputeLabel(directoryObject, objectIdentifier, domain, utils);
+
+            directoryObject.TryGetProperty(LDAPProperties.SAMAccountName, out var samAccountName);
+            res.DisplayName = ComputeDisplayName(directoryObject, domain, res.ObjectType, samAccountName);
```
Comment (definitelynotagoblin) [inline src/CommonLib/LdapUtils.cs:1210]: I would be careful that this returns an expected DisplayName when samAccountName is `null`

Comment (definitelynotagoblin) [reply src/CommonLib/LdapUtils.cs:1210]: If objecttype is `Base`, this may return something like "@[domain]"

Comment (rvazarkar) [reply src/CommonLib/LdapUtils.cs:1210]: Added some fallbacks

File: src/CommonLib/Ntlm/HttpNtlmAuthenticationService.cs [hunk 32]
```diff
@@ -0,0 +1,167 @@
...
+        _logger = logger ?? Logging.LogProvider.CreateLogger(nameof(HttpNtlmAuthenticationService));
+        _httpClientFactory = httpClientFactory;
+    }
+
+    public async Task EnsureRequiresAuth(Uri url, bool? useBadChannelBindings) {
+        if (url == null)
+            throw new ArgumentException("Url property is null");
+
+        if (useBadChannelBindings == null && url.Scheme == "https")
+            throw new ArgumentException("When using HTTPS, useBadChannelBindings must be set");
```
Comment (definitelynotagoblin) [inline src/CommonLib/Ntlm/HttpNtlmAuthenticationService.cs:30]: Nitpicking, but trinary args like this tend to create some funky logic flows that can be hard to follow if you're smooth-brained like I am.
Here I think we're taking on some complexity by managing arg validations that the callers could be doing instead.

I might split this into two functions for clarity: the public function that does everything up to this point, then a private `EnsureRequiresAuth(Uri url, bool useBadChannelBindings)` that this one calls, feeding the appropriate coerced `useBadChannelBindings`:

`EnsureRequiresAuth(url, useBadChannelBindings ?? true);`

File: src/CommonLib/Ntlm/HttpNtlmAuthenticationService.cs [hunk 33]
```diff
@@ -0,0 +1,167 @@
...
+        var schemes = response.Headers.WwwAuthenticate
+            .Select(header => header.Scheme)
+            .Where(scheme => scheme == "NTLM" || scheme == "Negotiate")
+            .Distinct()
+            .ToArray();
+
+        return schemes;
+    }
+
+    private async Task AuthWithBadChannelBindings(Uri url, string authScheme) {
```
Comment (definitelynotagoblin) [inline src/CommonLib/Ntlm/HttpNtlmAuthenticationService.cs]: Could use "-Async" suffix

File: src/CommonLib/OutputTypes/APIResult/APIResult.cs [hunk 34]
```diff
@@ -0,0 +1,29 @@
...
+        }
+
+        public static APIResult<T> Failure(string failureReason) {
+            return new APIResult<T> {
+                Collected = false,
+                FailureReason = failureReason
+            };
+        }
+        
+        public static implicit operator APIResult<T>(T input)
```
Comment (definitelynotagoblin) [inline src/CommonLib/OutputTypes/APIResult/APIResult.cs]: Today I learned implicit operators

File: src/CommonLib/OutputTypes/APIResult/AceRegistryAPIResult.cs [hunk 35]
```diff
@@ -2,7 +2,7 @@
 
 namespace SharpHoundCommonLib.OutputTypes
 {
-    public class AceRegistryAPIResult : APIResult
+    public class AceRegistryAPIResult : APIResult.APIResult
```
Comment (definitelynotagoblin) [inline src/CommonLib/OutputTypes/APIResult/AceRegistryAPIResult.cs]: Prefer the namespace and class had different names but not a blocker

File: src/CommonLib/OutputTypes/Computer.cs [hunk 36]
```diff
@@ -1,37 +1,76 @@
...
     /// </summary>
-    public class Computer : OutputBase
-    {
+    public class Computer : OutputBase {
         public string PrimaryGroupSID { get; set; }
-        public TypedPrincipal[] AllowedToDelegate { get; set; } = Array.Empty<TypedPrincipal>();
-        public TypedPrincipal[] AllowedToAct { get; set; } = Array.Empty<TypedPrincipal>();
-        public TypedPrincipal[] HasSIDHistory { get; set; } = Array.Empty<TypedPrincipal>();
-        public TypedPrincipal[] DumpSMSAPassword { get; set; } = Array.Empty<TypedPrincipal>();
+        public TypedPrincipal[] AllowedToDelegate { get; set; } = [];
```
Comment (definitelynotagoblin) [inline src/CommonLib/OutputTypes/Computer.cs:10]: This syntax may not compile with older dotnet sdks

Comment (definitelynotagoblin) [reply src/CommonLib/OutputTypes/Computer.cs:10]: Looks like this syntax feature isn't available til dotnet 8
`Feature 'collection expressions' is not available in C# 11.0. Please use language version 12.0 or greater.`

Comment (definitelynotagoblin) [reply src/CommonLib/OutputTypes/Computer.cs:10]: Thanks dotnet fiddle

**Review by definitelynotagoblin — CHANGES_REQUESTED** [review]
https://github.com/SpecterOps/SharpHoundCommon/pull/177/files#r1945552460

**Review by definitelynotagoblin — APPROVED** [review]

File: src/CommonLib/Enums/CollectionMethod.cs [hunk 37]
```diff
@@ -25,10 +23,19 @@ public enum CollectionMethod
         CARegistry = 1 << 16,
         DCRegistry = 1 << 17,
         CertServices = 1 << 18,
+        LdapServices = 1 << 19,
+        WebClientService = 1 << 21,
+        SmbInfo = 1 << 22,
+        //TODO: Re-introduce this when we're ready for Event Log collection
+        //EventLogs = 1 << 23,
```
Comment (ktstrader) [inline src/CommonLib/Enums/CollectionMethod.cs:30]: Looks like EventLogs should be uncommented because it is used in the ntlm updates for SharpHound [here](https://github.com/SpecterOps/SharpHound/pull/133/files#diff-81351f3d92bff48f24fe7b072073d3639d382eb32da99bfdbc200e6807fc2e99R210)

Comment (rvazarkar) [reply src/CommonLib/Enums/CollectionMethod.cs:30]: SH is not doing event log collection yet, so remove the reference on that side as well

File: src/CommonLib/Processors/CertAbuseProcessor.cs [hunk 38]
```diff
@@ -245,32 +237,26 @@ private RegistryResult GetEnrollmentAgentRights(string target, string caName)
...
             {
-                ret.FailureReason = data.FailureReason;
-                return ret;
+                return APIResult<bool>.Failure(data.FailureReason);
             }
 
             if (data.Value == null)
             {
-                return ret;
+                return APIResult<bool>.Success(false);
```
Comment (ktstrader) [inline src/CommonLib/Processors/CertAbuseProcessor.cs:255]: @leechristensen Changing type BoolRegistryAPIResult to APIResult<bool>, when data.Value is null I did have to set the return value to false. Is this an accurate value?

Comment (leechristensen) [reply src/CommonLib/Processors/CertAbuseProcessor.cs:255]: I believe that's correct, but I'll defer to @JonasBK since he's more familiar with this.

Comment (JonasBK) [reply src/CommonLib/Processors/CertAbuseProcessor.cs:255]: That is correct! 👍
